peers.py

```diff
@@ -42,3 +44,44 @@ def get_nodes(self):
             nodes.append((conv.scope, conv.get_remote('private-address')))
 
         return nodes
+
+    def restarted_nodes(self):
+        nodes = []
+        nonce = leader_get('restart_nonce')
+        if not nonce:
+            return nodes  # We're not restarting if no nonce is set.
+        for conv in self.conversations():
+            if conv.get_remote('restarted.{}'.format(nonce)):
+                nodes.append((conv.scope, conv.get_remote('private-address')))
+
+        return nodes
+
+    def set_zk_leader(self):
+        '''
+        Inform peers that the unit that calls this method is the Zookeeper leader.
+
+        Note that Zookeeper tracks leadership separately from juju;
+        the Zookeeper leader is not necessarily the Juju leader.
+
+        '''
+        for conv in self.conversations():
```

Comment:
If the leader changes because the leader has gone away, we trigger a new rolling restart, with a fresh queue, and everything works (I've tested it manually, though I haven't figured out how to write a formal test for it).

That's what all the nonce stuff and departed hooks are about: handling the situation where the leader changes.